### PR TITLE
[SYCL] Limit bfloat16 operators to scalar datatypes convertible to float

### PR DESCRIPTION
**Pull request by fineg74**
Aligns implementation with the spec: https://github.com/intel/llvm/blob/48ec4dd6ff45b25b405b99d63f3c5b8537b1475d/sycl/doc/extensions/supported/sycl_ext_oneapi_bfloat16.asciidoc

### CODE REVIEW

**Review by v-klochkov — APPROVED** [review]
The patch looks better than what was before it.
I am not sure it will be sufficient for 1 element simd objects such as `simd<float, 1>`, which are convertible to `float`, but let's handle those cases in separate PRs.

The std::is_convertible make the implementation better, but not fully correct as the main/ideological issue still exists - bfloat16 implementation defines the behavior for user types without consent. If user wants `T` to be convertible to `float` and also re-define `operator+(T, bfloat16)`, then cannot efficiently do that without getting 'ambiguous call` error.

**Comment by lbushi25** [comment]
I also think its the approach that introduces the least additional complexity so looks good to me.

**Review by lbushi25 — APPROVED** [review]

File: sycl/include/sycl/ext/oneapi/bfloat16.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/bfloat16.hpp:166]: The current version of the extension returns `bfloat16` for both.

File: sycl/test-e2e/ESIMD/regression/bfloat16_vector_plus_scalar_pvc.cpp
Comment (aelovikov-intel) [inline sycl/test-e2e/ESIMD/regression/bfloat16_vector_plus_scalar_pvc.cpp:8]: Please add a comment why.

File: sycl/include/sycl/ext/oneapi/bfloat16.hpp
Comment (v-klochkov) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:166]: This function cannot return 'bfloat16', it returns the modified `lhs`, which is of `T` type, e.g.: `T lhs; lhs += bfloat_var;'

File: sycl/include/sycl/ext/oneapi/bfloat16.hpp
Comment (v-klochkov) [inline sycl/include/sycl/ext/oneapi/bfloat16.hpp:184]: Not about this PR, but rather about the approach used for `bfloat16`. IMO this line is one more proof of using wrong approach for `bfloat16` comparing to `sycl::half`.

Look, here the return type of operator+ is always `bfloat16`, while it should be determined by C++ type-promotion-rules usually choosing a heavier type, e.g. 'double` + `float` results into `double`.

Here: `double` + `bfloat16` is `bfloat16`, which is wrong IMO.

File: sycl/include/sycl/ext/oneapi/bfloat16.hpp
Comment (aelovikov-intel) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:166]: \+ @gmlueck

Comment (v-klochkov) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:166]: @aelovikov-intel - this patch does not resolve the issue mentioned in this mini-discussion. It leaves the result type as it was before this patch as it is focused on different aspect.

The return type indeed should be fixed for operator+, not for operator+= where `lhs` is of `T` type, but that is a topic for a separate discussion & pull-request.

Comment (aelovikov-intel) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:166]: There is no guarantee `lhs = f;` would work, is it?

Comment (v-klochkov) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:166]: @aelovikov-intel I don't understand, can you elaborate your comment? What do you suggest to do in this mini-discussion?

Comment (aelovikov-intel) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:166]: I don't think we should be working on this patch before the spec is fixed. If @gmlueck is on vacation, then the PR author should work with his backup to get the spec fixed first.

Comment (v-klochkov) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:166]: We have already agreed with @gmlueck that std::is_convertible is what is meant in the existing SPEC.

Let's not wait until the ALL issues in spec resolved, it will take long and ESIMD will stay blocked by the wrong implementation.

Comment (v-klochkov) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:166]: @steffenlarsen - can you please also review the patch as @gmlueck backup (as written in Greg's OOO).

Comment (steffenlarsen) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:166]: I don't see these operators in the [sycl_ext_oneapi_bfloat16 extension](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_bfloat16.asciidoc). I only see the one with two bfloat16 operands. As such, I don't think they should be here unless we add them to the extension. Since they were here already and the extension is considered supported, we cannot just remove them, so we will have to consider deprecating them (or adding them to the extension.)

Comment (v-klochkov) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:166]: @steffenlarsen - the spec has such lines:

> template <typename T>
>   friend bfloat16 operatorOP(const T &lhs, const bfloat16 &rhs) { /* …​ */ }
> 
> Construct a new instance of the bfloat16 class with the value of the new bfloat16 instance being the result of an OP arithmetic operation between the lhs of template type T and rhs bfloat16 value. Type T must be convertible to float.
> 
> OP is +, -, *, /

This patch here only syncs the implementation with this SPEC statement: "Type T must be convertible to float."

The attempt to 'static_cast<float>(lhs), where lhs is not convertible to float is simply a bug in implementation, fixed here.

**Comment by v-klochkov** [comment]
> 1. Could we consider making it a `static_assert` in the operator defintions instead of using SFINAE? If we are not planning on offering alternative definitions, this might give the user a better error message when they misuse it.

@steffenlarsen - this is not a usable option. simd<float, 8> is not convertible to 'bfloat16', and we have another `operator+` for `simd<float,8>` and `bfloat16` defined in ESIMD headers, but the operator+ defined in `bfloat16.hpp` also considered as option now and would give `static_assert` proposed in (1) above, which does not make any sense for `simd<float, 8` because the bfloat16.hpp::operator+ simply should not be considered as operator for `simd<float, 8>` and should be filtered away by SFINAE.

File: sycl/include/sycl/ext/oneapi/bfloat16.hpp
Comment (steffenlarsen) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:166]: I agree with you, the changes in the patch adheres to the convertibility requirement in the extension, even though the operators here are technically not part of the extension. My primary point above is in response to the initial comment of

> The current version of the extension returns bfloat16 for both.

which for the operators right here aren't applicable as they are not even in the spec. The non-assignment binary operators below seem to have the correct types in accordance with the spec.

**Comment by v-klochkov** [comment]
> 2\. This is not a blocker, but I wonder why we went with the wording in the spec to require T to be convertible to float, when we could have just made the argument a float and relied on implicit conversion. Ambiguity problems maybe? @aelovikov-intel - Can you think of anything?

@steffenlarsen - that is what @fineg74 did in the commit (https://github.com/intel/llvm/pull/12089 + 1 minor extra fix ( https://github.com/intel/llvm/commit/c8ebf12a39a20f1efd22919a5e73a100d105b72b ) making `bfloat16` working exactly the same way as `sycl::half`, but had to revert it later because it was considered as breaking backward compatibility in some corner cases, e.g. when type `T` is implicitly convertible to `float` and `int`.

More about the SPEC issue:
llvm/intel issue: https://github.com/intel/llvm/issues/11244
Lots of discussions without conclusion in this PR: https://github.com/intel/llvm/pull/12327

**Comment by steffenlarsen** [comment]
Thank you for clarifying, @v-klochkov ! I think https://github.com/intel/llvm/pull/12089 overdid it slightly by adding more definitions than just one with `float`, then all calls with arguments convertible to `float` would rely on implicit conversion. However, given the spec has a template argument, we can't do it without a change to the spec first. Let's discuss it in https://github.com/intel/llvm/issues/11244.

I will leave approval of this to @aelovikov-intel, but to me it looks good.

**Comment by aelovikov-intel** [comment]
> I will leave approval of this to @aelovikov-intel, but to me it looks good.

Everybody thinks the extension itself has to be changed, yet this PR tries to modify implementation without doing that (and align with the unmodified spec). Doesn't sound right to me. What do I miss?

**Comment by v-klochkov** [comment]
> > I will leave approval of this to @aelovikov-intel, but to me it looks good.
> 
> Everybody thinks the extension itself has to be changed, yet this PR tries to modify implementation without doing that (and align with the unmodified spec). Doesn't sound right to me. What do I miss?

This PR only brings the implementation closer to SPEC and the motivation is to unblock the bfloat16 for `esimd::simd`.
There is no even 1 case (corner or basic) that would be broken as result of this PR. But it unblocks `bfloat16` for `esimd::simd`.

Discussing/changing the SPEC and implementation will take long or even forever (taking into account that "good" solution for SPEC issues may result into the backward-compatibility for corner cases). 
All that time ESIMD would be blocked by implementational issue being fixed here.

File: sycl/include/sycl/ext/oneapi/bfloat16.hpp
Comment (aelovikov-intel) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:166]: I don't want to approve a PR that does not lessen the amount of technical debt in the codebase just because doing the right thing "will take long or even forever".

I can only assume that "forever" would happen if everybody would try to hack their way around (like this PR), instead of just doing the work on the specification.

It's not even the core SYCL, it's the extension that is part of this repository...

To clarify, I'm not going to prevent this merged if you find somebody else who'll give the approval, but it won't be me.

**Comment by steffenlarsen** [comment]
> Discussing/changing the SPEC and implementation will take long or even forever (taking into account that "good" solution for SPEC issues may result into the backward-compatibility for corner cases).

I think this is an important point to stress. Even if we fix the extension, it is considered "supported" and not "experimental", so we have to take care about how and when we break it. As such, we should strive to fix the implementation in accordance with the current specification in the meantime.

File: sycl/include/sycl/ext/oneapi/bfloat16.hpp
Comment (steffenlarsen) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:166]: Please see https://github.com/intel/llvm/pull/12477#issuecomment-1918713047. I agree there are things that need fixing in the specification, but these changes are in accordance with the current specification and are as such a bug-fix. Addressing additional problems in the extension cannot be made to the implementation immediately, as this isn't an experimental extension.

These operators are in a gray area, as they are not part of the extension, but since we made the mistake of exposing them, we should give them the same treatment we did with the extension interfaces, lest we deprecate them.

Comment (aelovikov-intel) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:166]: > These operators are in a gray area, as they are not part of the extension, but since we made the mistake of exposing them, we should give them the same treatment we did with the extension interfaces, or deprecate them.

Then

1) It makes all the sense to deprecate them here
2) SFINAE enabler is wrong for those gray operators - `lhs = f` is not guaranteed to work. And the argument that the spec requires different requirements doesn't apply here (because I could use the same argument to remove them immediately).
3) I think it's debatable if a bugfix like this is even considered breaking change.

File: sycl/include/sycl/ext/intel/esimd/detail/bfloat16_type_traits.hpp
Comment (rolandschulz) [inline sycl/include/sycl/ext/intel/esimd/detail/bfloat16_type_traits.hpp:97]: this isn't used?

**Review by gmlueck — APPROVED** [review]

**Review by aelovikov-intel — APPROVED** [review]

**Pull request merged**
